// ngx-drag-to-select/esm5/lib/operators.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { filter, map, withLatestFrom, distinctUntilChanged } from 'rxjs/operators';
import { getRelativeMousePosition, hasMinimumSize } from './utils';
/** @type {?} */
export var createSelectBox = function (container) { return function (source) {
    return source.pipe(map(function (_a) {
        var _b = tslib_1.__read(_a, 3), event = _b[0], opacity = _b[1], _c = _b[2], x = _c.x, y = _c.y;
        // Type annotation is required here, because `getRelativeMousePosition` returns a `MousePosition`,
        // the TS compiler cannot figure out the shape of this type.
        /** @type {?} */
        var mousePosition = getRelativeMousePosition(event, container);
        /** @type {?} */
        var width = opacity > 0 ? mousePosition.x - x : 0;
        /** @type {?} */
        var height = opacity > 0 ? mousePosition.y - y : 0;
        return {
            top: height < 0 ? mousePosition.y : y,
            left: width < 0 ? mousePosition.x : x,
            width: Math.abs(width),
            height: Math.abs(height),
            opacity: opacity
        };
    }));
}; };
/** @type {?} */
export var whenSelectBoxVisible = function (selectBox$) { return function (source) {
    return source.pipe(withLatestFrom(selectBox$), filter(function (_a) {
        var _b = tslib_1.__read(_a, 2), selectBox = _b[1];
        return hasMinimumSize(selectBox, 0, 0);
    }), map(function (_a) {
        var _b = tslib_1.__read(_a, 2), event = _b[0], _ = _b[1];
        return event;
    }));
}; };
/** @type {?} */
export var distinctKeyEvents = function () { return function (source) {
    return source.pipe(distinctUntilChanged(function (prev, curr) {
        return prev.keyCode === curr.keyCode;
    }));
}; };
