// ngx-drag-to-select/esm5/lib/shortcut.service.js
import * as tslib_1 from "tslib";
var _a;
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingReturn,uselessCode} checked by tsc
 */
import { Injectable, Inject } from '@angular/core';
import { CONFIG } from './tokens';
/** @type {?} */
var SUPPORTED_KEYS = {
    alt: true,
    shift: true,
    meta: true,
    ctrl: true
};
/** @type {?} */
var META_KEY = 'meta';
/** @type {?} */
var KEY_ALIASES = (_a = {},
    _a[META_KEY] = ['ctrl', 'meta'],
    _a);
/** @type {?} */
var SUPPORTED_SHORTCUTS = {
    disableSelection: true,
    toggleSingleItem: true,
    addToSelection: true,
    removeFromSelection: true
};
/** @type {?} */
var ERROR_PREFIX = '[ShortcutService]';
var ShortcutService = /** @class */ (function () {
    function ShortcutService(config) {
        this.config = config;
        this._shortcuts = {};
        this._shortcuts = this.createShortcutsFromConfig(config.shortcuts);
    }
    /**
     * @param {?} event
     * @return {?}
     */
    ShortcutService.prototype.disableSelection = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        return this.isShortcutPressed('disableSelection', event);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    ShortcutService.prototype.toggleSingleItem = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        return this.isShortcutPressed('toggleSingleItem', event);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    ShortcutService.prototype.addToSelection = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        return this.isShortcutPressed('addToSelection', event);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    ShortcutService.prototype.removeFromSelection = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        return this.isShortcutPressed('removeFromSelection', event);
    };
    /**
     * @param {?} event
     * @return {?}
     */
    ShortcutService.prototype.extendedSelectionShortcut = /**
     * @param {?} event
     * @return {?}
     */
    function (event) {
        return this.addToSelection(event) || this.removeFromSelection(event);
    };
    /**
     * @param {?} shortcuts
     * @return {?}
     */
    ShortcutService.prototype.createShortcutsFromConfig = /**
     * @param {?} shortcuts
     * @return {?}
     */
    function (shortcuts) {
        var _this = this;
        var e_1, _a;
        /** @type {?} */
        var shortcutMap = {};
        var _loop_1 = function (key, shortcutsForCommand) {
            if (!this_1.isSupportedShortcut(key)) {
                throw new Error(this_1.getErrorMessage("Shortcut " + key + " not supported"));
            }
            shortcutsForCommand
                .replace(/ /g, '')
                .split(',')
                .forEach(function (shortcut) {
                if (!shortcutMap[key]) {
                    shortcutMap[key] = [];
                }
                /** @type {?} */
                var combo = shortcut.split('+');
                /** @type {?} */
                var cleanCombos = _this.substituteKey(shortcut, combo, META_KEY);
                cleanCombos.forEach(function (cleanCombo) {
                    /** @type {?} */
                    var unsupportedKey = _this.isSupportedCombo(cleanCombo);
                    if (unsupportedKey) {
                        throw new Error(_this.getErrorMessage("Key '" + unsupportedKey + "' in shortcut " + shortcut + " not supported"));
                    }
                    shortcutMap[key].push(cleanCombo.map(function (comboKey) { return comboKey + "Key"; }));
                });
            });
        };
        var this_1 = this;
        try {
            for (var _b = tslib_1.__values(Object.entries(shortcuts)), _c = _b.next(); !_c.done; _c = _b.next()) {
                var _d = tslib_1.__read(_c.value, 2), key = _d[0], shortcutsForCommand = _d[1];
                _loop_1(key, shortcutsForCommand);
            }
        }
        catch (e_1_1) { e_1 = { error: e_1_1 }; }
        finally {
            try {
                if (_c && !_c.done && (_a = _b.return)) _a.call(_b);
            }
            finally { if (e_1) throw e_1.error; }
        }
        return shortcutMap;
    };
    /**
     * @param {?} shortcut
     * @param {?} combo
     * @param {?} substituteKey
     * @return {?}
     */
    ShortcutService.prototype.substituteKey = /**
     * @param {?} shortcut
     * @param {?} combo
     * @param {?} substituteKey
     * @return {?}
     */
    function (shortcut, combo, substituteKey) {
        /** @type {?} */
        var hasSpecialKey = shortcut.includes(substituteKey);
        /** @type {?} */
        var substitutedShortcut = [];
        if (hasSpecialKey) {
            /** @type {?} */
            var cleanShortcut_1 = combo.filter(function (element) { return element !== META_KEY; });
            KEY_ALIASES.meta.forEach(function (alias) {
                substitutedShortcut.push(tslib_1.__spread(cleanShortcut_1, [alias]));
            });
        }
        else {
            substitutedShortcut.push(combo);
        }
        return substitutedShortcut;
    };
    /**
     * @param {?} message
     * @return {?}
     */
    ShortcutService.prototype.getErrorMessage = /**
     * @param {?} message
     * @return {?}
     */
    function (message) {
        return ERROR_PREFIX + " " + message;
    };
    /**
     * @param {?} shortcutName
     * @param {?} event
     * @return {?}
     */
    ShortcutService.prototype.isShortcutPressed = /**
     * @param {?} shortcutName
     * @param {?} event
     * @return {?}
     */
    function (shortcutName, event) {
        /** @type {?} */
        var shortcuts = this._shortcuts[shortcutName];
        return shortcuts.some(function (shortcut) {
            return shortcut.every(function (key) { return event[key]; });
        });
    };
    /**
     * @param {?} combo
     * @return {?}
     */
    ShortcutService.prototype.isSupportedCombo = /**
     * @param {?} combo
     * @return {?}
     */
    function (combo) {
        /** @type {?} */
        var unsupportedKey = null;
        combo.forEach(function (key) {
            if (!SUPPORTED_KEYS[key]) {
                unsupportedKey = key;
                return;
            }
        });
        return unsupportedKey;
    };
    /**
     * @param {?} shortcut
     * @return {?}
     */
    ShortcutService.prototype.isSupportedShortcut = /**
     * @param {?} shortcut
     * @return {?}
     */
    function (shortcut) {
        return SUPPORTED_SHORTCUTS[shortcut];
    };
    ShortcutService.decorators = [
        { type: Injectable }
    ];
    /** @nocollapse */
    ShortcutService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [CONFIG,] }] }
    ]; };
    return ShortcutService;
}());
export { ShortcutService };
if (false) {
    /** @type {?} */
    ShortcutService.prototype._shortcuts;
    /** @type {?} */
    ShortcutService.prototype.config;
}
